File: nativelib/src/main/scala/scala/scalanative/posix/netinet/in.scala
package scala.scalanative
package posix
package netinet

import scalanative.native._
import scalanative.posix.inttypes._
import scalanative.posix.sys.socket

@extern
object in {
  type in_port_t = uint16_t
  type in_addr_t = uint32_t
  type _16       = Nat.Digit[Nat._1, Nat._6]

  type in_addr = CStruct1[in_addr_t] // s_addr
  type sockaddr_in = CStruct3[socket.sa_family_t, // sin_family
                              in_port_t, // sin_port
                              in_addr] // sin_addr

  type in6_addr = CStruct1[CArray[uint8_t, _16]] // s6_addr
  type sockaddr_in6 = CStruct5[in6_addr, // sin6_addr
                               socket.sa_family_t, // sin6_family
                               in_port_t, // sin6_port
                               uint32_t, // sin6_flowinfo
                               uint32_t] // sin6_scope_id

  type ipv6_mreq = CStruct2[in6_addr, // ipv6mr_multiaddr
                            CUnsignedInt] // ipv6mr_interface

  @name("scalanative_IPROTO_IP")
  def IPROTO_IP: CInt = extern

  @name("scalanative_IPPROTO_IPV6")
  def IPPROTO_IPV6: CInt = extern

  @name("scalanative_IPPROTO_ICMP")
  def IPPROTO_ICMP: CInt = extern

  @name("scalanative_IPPROTO_RAW")
  def IPPROTO_RAW: CInt = extern

  @name("scalanative_IPPROTO_TCP")
  def IPPROTO_TCP: CInt = extern

  @name("scalanative_IPPROTO_UDP")
  def IPPROTO_UDP: CInt = extern

  @name("scalanative_INADDR_ANY")
  def INADDR_ANY: uint32_t = extern

  @name("scalanative_INADDR_BROADCAST")
  def INADDR_BROADCAST: uint32_t = extern

  @name("scalanative_INET6_ADDRSTRLEN")
  def INET6_ADDRSTRLEN: CInt = extern

  @name("scalanative_INET_ADDRSTRLEN")
  def INET_ADDRSTRLEN: CInt = extern

  @name("scalanative_IPV6_JOIN_GROUP")
  def IPV6_JOIN_GROUP: CInt = extern

  @name("scalanative_IPV6_LEAVE_GROUP")
  def IPV6_LEAVE_GROUP: CInt = extern

  @name("scalanative_IPV6_MULTICAST_HOPS")
  def IPV6_MULTICAST_HOPS: CInt = extern

  @name("scalanative_IPV6_MULTICAST_IF")
  def IPV6_MULTICAST_IF: CInt = extern

  @name("scalanative_IPV6_MULTICAST_LOOP")
  def IPV6_MULTICAST_LOOP: CInt = extern

  @name("scalanative_IPV6_UNICAST_HOPS")
  def IPV6_UNICAST_HOPS: CInt = extern

  @name("scalanative_IPV6_V6ONLY")
  def IPV6_V6ONLY: CInt = extern

  @name("scalanative_IN6_IS_ADDR_UNSPECIFIED")
  def IN6_IS_ADDR_UNSPECIFIED(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_LOOPBACK")
  def IN6_IS_ADDR_LOOPBACK(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MULTICAST")
  def IN6_IS_ADDR_MULTICAST(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_LINKLOCAL")
  def IN6_IS_ADDR_LINKLOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_SITELOCAL")
  def IN6_IS_ADDR_SITELOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_V4MAPPED")
  def IN6_IS_ADDR_V4MAPPED(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_V4COMPAT")
  def IN6_IS_ADDR_V4COMPAT(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MC_NODELOCAL")
  def IN6_IS_ADDR_MC_NODELOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MC_LINKLOCAL")
  def IN6_IS_ADDR_MC_LINKLOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MC_SITELOCAL")
  def IN6_IS_ADDR_MC_SITELOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MC_ORGLOCAL")
  def IN6_IS_ADDR_MC_ORGLOCAL(arg: Ptr[in6_addr]): CInt = extern

  @name("scalanative_IN6_IS_ADDR_MC_GLOBAL")
  def IN6_IS_ADDR_MC_GLOBAL(arg: Ptr[in6_addr]): CInt = extern

}

object inOps {
  import in._

  implicit class sockaddr_inOps(val ptr: Ptr[sockaddr_in]) extends AnyVal {
    def sin_family: socket.sa_family_t = !ptr._1
    def sin_port: in_port_t            = !ptr._2
    def sin_addr: Ptr[in_addr]         = ptr._3

    def sin_family_=(v: socket.sa_family_t): Unit = !ptr._1 = v
    def sin_port_=(v: in_port_t): Unit            = !ptr._2 = v
    def sin_addr_=(v: Ptr[in_addr]): Unit         = !ptr._3 = !v
  }

  implicit class sockaddr_in6Ops(val ptr: Ptr[sockaddr_in6]) extends AnyVal {
    def sin6_addr: Ptr[in6_addr]        = ptr._1
    def sin6_family: socket.sa_family_t = !ptr._2
    def sin6_port: in_port_t            = !ptr._3
    def sin6_flowinfo: uint32_t         = !ptr._4
    def sin6_scope_id: uint32_t         = !ptr._5

    def sin6_addr_=(v: Ptr[in6_addr]): Unit        = !ptr._1 = !v
    def sin6_family_=(v: socket.sa_family_t): Unit = !ptr._2 = v
    def sin6_port_=(v: in_port_t): Unit            = !ptr._3 = v
    def sin6_flowinfo_=(v: uint32_t): Unit         = !ptr._4 = v
    def sin6_scope_id_=(v: uint32_t): Unit         = !ptr._5 = v
  }

  implicit class in_addrOps(val ptr: Ptr[in_addr]) extends AnyVal {
    def in_addr: in_addr_t            = !ptr._1
    def in_addr_=(v: in_addr_t): Unit = !ptr._1 = v
  }

  implicit class in6_addrOps(val ptr: Ptr[in6_addr]) extends AnyVal {
    def s6_addr: CArray[uint8_t, _16]            = !ptr._1
    def s6_addr_=(v: CArray[uint8_t, _16]): Unit = !ptr._1 = v
  }

  implicit class ipv6_mreqOps(val ptr: Ptr[ipv6_mreq]) extends AnyVal {
    def ipv6mr_multiaddr: Ptr[in6_addr] = ptr._1
    def ipv6mr_interface: CUnsignedInt  = !ptr._2

    def ipv6mr_multiaddr_=(v: Ptr[in6_addr]): Unit = !ptr._1 = !v
    def ipv6mr_interface_=(v: CUnsignedInt): Unit  = !ptr._2 = v
  }
}

File: posixlib/src/main/resources/scala-native/netinet/in.c
#if defined(SCALANATIVE_COMPILE_ALWAYS) ||                                     \
    defined(__SCALANATIVE_POSIX_NETINET_IN)
#include <stddef.h>
#include <string.h>
#include "in.h"

#if !(defined __STDC_VERSION__) || (__STDC_VERSION__ < 201112L)
#ifndef SCALANATIVE_SUPPRESS_STRUCT_CHECK_WARNING
#warning "Size and order of C structures are not checked when -std < c11."
#endif
#else
// Posix defines the name and type of required fields. Size of fields
// and any internal or tail padding are left unspecified. This section
// verifies that the C and Scala Native definitions match in each compilation
// environment.

// IPv4
_Static_assert(sizeof(struct scalanative_sockaddr_in) == 16,
               "Unexpected size: scalanative_sockaddr_in");

_Static_assert(sizeof(struct scalanative_sockaddr_in) ==
                   sizeof(struct sockaddr_in),
               "Unexpected size: os sockaddr_in");

// On systems which define/use IETF RFC SIN6_LEN macro, sin_family &
// sin_len are synthesized fields, managed by Ops access routines in in.scala.
// C offsetof() sin_family will be 2 for the OS sockaddr_in, but strictly 0 for
// scalanative_sockaddr_in. Scala access routines will access the
// expected bytes.

_Static_assert(offsetof(struct scalanative_sockaddr_in, sin_family) == 0,
               "Unexpected offset: scalanative_sockaddr_in.sin_family");

_Static_assert(offsetof(struct scalanative_sockaddr_in, sin_port) ==
                   offsetof(struct sockaddr_in, sin_port),
               "Unexpected offset: sockaddr_in.sin_port");

_Static_assert(offsetof(struct scalanative_sockaddr_in, sin_addr) ==
                   offsetof(struct sockaddr_in, sin_addr),
               "Unexpected offset: sockaddr_in.sin_addr");
// IPv6
_Static_assert(sizeof(struct scalanative_sockaddr_in6) == 28,
               "Unexpected size: scalanative_sockaddr_in6");

_Static_assert(sizeof(struct scalanative_sockaddr_in6) ==
                   sizeof(struct sockaddr_in6),
               "Unexpected size: os sockaddr_in");

// For systems which define/use IETF RFC SIN6_LEN macro, sin6_family &
// sin6_len see comment above for corresponding scalanative_sockaddr_in6.

_Static_assert(offsetof(struct scalanative_sockaddr_in6, sin6_family) == 0,
               "Unexpected offset: scalanative_sockaddr_in6.sin6_family");

_Static_assert(offsetof(struct scalanative_sockaddr_in6, sin6_port) ==
                   offsetof(struct sockaddr_in6, sin6_port),
               "Unexpected offset: sockaddr_in6.sin6_port");

_Static_assert(offsetof(struct scalanative_sockaddr_in6, sin6_flowinfo) ==
                   offsetof(struct sockaddr_in6, sin6_flowinfo),
               "Unexpected offset: sockaddr_in6.sin6_flowinfo");

_Static_assert(offsetof(struct scalanative_sockaddr_in6, sin6_addr) ==
                   offsetof(struct sockaddr_in6, sin6_addr),
               "Unexpected offset: sockaddr_in6.sin6_addr");

_Static_assert(offsetof(struct scalanative_sockaddr_in6, sin6_scope_id) ==
                   offsetof(struct sockaddr_in6, sin6_scope_id),
               "Unexpected offset: sockaddr_in6.sin6_scope_id");

#endif // structure checking

int scalanative_ipproto_ip() { return IPPROTO_IP; }

int scalanative_ipproto_ipv6() { return IPPROTO_IPV6; }

int scalanative_ipproto_icmp() { return IPPROTO_ICMP; }

int scalanative_ipproto_raw() { return IPPROTO_RAW; }

int scalanative_ipproto_tcp() { return IPPROTO_TCP; }

int scalanative_ipproto_udp() { return IPPROTO_UDP; }

uint32_t scalanative_inaddr_any() { return INADDR_ANY; }

uint32_t scalanative_inaddr_broadcast() { return INADDR_BROADCAST; }

int scalanative_inet6_addrstrlen() { return INET6_ADDRSTRLEN; }

int scalanative_inet_addrstrlen() { return INET_ADDRSTRLEN; }

int scalanative_ipv6_join_group() { return IPV6_JOIN_GROUP; }

int scalanative_ipv6_leave_group() { return IPV6_LEAVE_GROUP; }

int scalanative_ipv6_multicast_hops() { return IPV6_MULTICAST_HOPS; }

int scalanative_ipv6_multicast_if() { return IPV6_MULTICAST_IF; }

int scalanative_ipv6_multicast_loop() { return IPV6_MULTICAST_LOOP; }

int scalanative_ipv6_unicast_hops() { return IPV6_UNICAST_HOPS; }

int scalanative_ipv6_v6only() { return IPV6_V6ONLY; }

int scalanative_ip_multicast_if() { return IP_MULTICAST_IF; }

int scalanative_ip_multicast_loop() { return IP_MULTICAST_LOOP; }

int scalanative_ip_tos() { return IP_TOS; }

int scalanative_in6_is_addr_unspecified(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_UNSPECIFIED((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_loopback(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_LOOPBACK((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_multicast(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MULTICAST((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_linklocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_LINKLOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_sitelocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_SITELOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_v4mapped(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_V4MAPPED((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_v4compat(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_V4COMPAT((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_mc_nodelocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MC_NODELOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_mc_linklocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MC_LINKLOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_mc_sitelocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MC_SITELOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_mc_orglocal(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MC_ORGLOCAL((struct in6_addr *)arg);
}

int scalanative_in6_is_addr_mc_global(struct scalanative_in6_addr *arg) {
    return IN6_IS_ADDR_MC_GLOBAL((struct in6_addr *)arg);
}
#endif